predicate/predicate.py

```diff
@@ -357,3 +361,99 @@ def get_values_list(obj, *lookups, **kwargs):
     else:
         return [tuple(value_dict[lookup].value for lookup in lookups)
                 for value_dict in value_dicts]
+
+
+class PredicateQuerySet(object):
+    """
+    Iterable wrapper that follows the QuerySet API.
+    """
+    def __init__(self, iterable, p=None):
+        self._evaluated = False
+        self.iterable = iterable
+        if p is None:
+            p = P()
+        self.P = p
+
+    def _evaluate(self):
+        """
+        Applies own filters to stored iterable.
+        """
+        if self._evaluated:
+            return
+        self._evaluated = True
+        self.iterable = self.P.filter(self.iterable)
+
+    def __repr__(self):
+        self._evaluate()
+        data = self.iterable[:REPR_OUTPUT_SIZE + 1]
+        if len(data) > REPR_OUTPUT_SIZE:
+            data[-1] = "...(remaining elements truncated)..."
+        return '<PredicateQuerySet %r>' % data
+
+    def __getitem__(self, k):
+        self._evaluate()
+        if isinstance(k, slice):
+            return self.__class__(self.iterable[k])
+        return self.iterable[k]
+
+    def __iter__(self):
+        return iter(self.iterable)
+
+    def _clone(self):
+        clone = type(self)(None)
+        for name, value in self.__dict__.iteritems():
+            setattr(clone, name, copy.deepcopy(value))
+        clone._evaluated = False
+        return clone
+
+    def all(self):
+        return self._clone()
+
+    def filter(self, *args, **kwargs):
+        clone = self._clone()
+        clone.P.children.append(P(*args, **kwargs))
+        return clone
+
+    def exclude(self, *args, **kwargs):
+        clone = self._clone()
+        clone.P.children.append(~P(*args, **kwargs))
+        return clone
+
+    def get(self, *args, **kwargs):
+        clone = self.filter(*args, **kwargs)
+        return clone.P.get(clone.iterable)
+
+    def exists(self):
+        self._evaluate()
+        return bool(self.iterable)
+
+    def count(self):
+        self._evaluate()
+        return len(self.iterable)
+
+    def _combine(self, other, connector):
+        clone1 = self._clone()
+        clone2 = other._clone()
+        iterable = itertools.chain(clone1.iterable, clone2.iterable)
```

Comment:
This is risky, since the iterable will be consumed the first time it's evaluated. Maybe cast to a list?